sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,711 @@
+      connectionFactory.setUri(uri());
+
+      connectionFactory.setAutomaticRecoveryEnabled(automaticRecovery());
+      connectionFactory.setConnectionTimeout(connectionTimeout());
+      connectionFactory.setNetworkRecoveryInterval(networkRecoveryInterval());
+      connectionFactory.setRequestedHeartbeat(requestedHeartbeat());
+      connectionFactory.setTopologyRecoveryEnabled(topologyRecovery());
+      connectionFactory.setRequestedChannelMax(requestedChannelMax());
+      connectionFactory.setRequestedFrameMax(requestedFrameMax());
+
+      return connectionFactory;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to consume messages from RabbitMQ server.
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<RabbitMqMessage>> {
+
+    @Nullable abstract ConnectionConfig connectionConfig();
+    @Nullable abstract String queue();
+    abstract boolean queueDeclare();
+    @Nullable abstract String exchange();
+    @Nullable abstract String exchangeType();
+    @Nullable abstract String routingKey();
+    @Nullable abstract Boolean useCorrelationId();
+    abstract long maxNumRecords();
+    @Nullable abstract Duration maxReadTime();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfig(ConnectionConfig connectionConfig);
+      abstract Builder setQueue(String queue);
+      abstract Builder setQueueDeclare(boolean queueDeclare);
+      abstract Builder setExchange(String exchange);
+      abstract Builder setExchangeType(String exchangeType);
+      abstract Builder setRoutingKey(String routingKey);
+      abstract Builder setUseCorrelationId(Boolean useCorrelationId);
+      abstract Builder setMaxNumRecords(long maxNumRecords);
+      abstract Builder setMaxReadTime(Duration maxReadTime);
+      abstract Read build();
+    }
+
+    public Read withConnectionConfig(ConnectionConfig connectionConfig) {
+      checkArgument(connectionConfig != null, "connectionConfig can not be null");
+      return builder().setConnectionConfig(connectionConfig).build();
+    }
+
+    public Read withUri(String uri) {
+      checkArgument(uri != null, "uri can not be null");
+      return builder().setConnectionConfig(connectionConfig().withUri(uri)).build();
+    }
+
+    public Read withQueue(String queue) {
+      checkArgument(queue != null, "queue can not be null");
+      return builder().setQueue(queue).build();
+    }
+
+    public Read withQueueDeclare(boolean queueDeclare) {
+      return builder().setQueueDeclare(queueDeclare).build();
+    }
+
+    public Read withExchange(String name, String type, String routingKey) {
+      checkArgument(name != null, "name can not be null");
+      checkArgument(type != null, "type can not be null");
+      checkArgument(routingKey != null, "routingKey can not be null");
+      checkArgument(type.equals("direct")
```

Comment:
Is it responsibility of RabbitMqIO to enforce this? What if we use this with a new version of RabbitMq that introduces a new exchange type?